src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs

```diff
@@ -382,26 +410,66 @@ public override int CountTokens(string sequence, bool isSpecialToken)
                 return null;
             }
 
-            List<byte> utf8Bytes = new();
-            bool useSpecialTokens = !skipSpecialTokens && _specialTokensDecoder is not null;
-
-            foreach (int id in ids)
+            byte[]? arrayPoolArray = null;
+            try
             {
-                if (_decoder.TryGetValue(id, out byte[]? tokenBytes))
+                Span<byte> utf8Bytes = stackalloc byte[256];
+                int utf8ByteCount = 0;
+
+                bool useSpecialTokens = !skipSpecialTokens && _specialTokensDecoder is not null;
+
+                foreach (int id in ids)
                 {
-                    utf8Bytes.AddRange(tokenBytes);
+                    if (_decoder.TryGetValue(id, out byte[]? tokenBytes))
+                    {
+                        if ((uint)utf8ByteCount + (uint)tokenBytes.Length > (uint)utf8Bytes.Length)
+                        {
+                            ArrayPoolGrow(ref utf8Bytes, ref arrayPoolArray, utf8ByteCount + tokenBytes.Length);
```

Comment:
> ArrayPoolGrow(ref utf8Bytes, ref arrayPoolArray, utf8ByteCount + tokenBytes.Length); [](http://example.com/codeflow?start=27&length=85)

should we grow the array with more spaces just in case more stuff we can add later without regrowing it?

Reply:
never mind, it is not going to grow more in this path.